flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/maxwell/MaxwellJsonDeserializationSchema.java

```diff
@@ -0,0 +1,181 @@
+import org.apache.flink.table.types.logical.RowType;
+import org.apache.flink.types.RowKind;
+import org.apache.flink.util.Collector;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static java.lang.String.format;
+import static org.apache.flink.table.types.utils.TypeConversions.fromLogicalToDataType;
+
+/**
+ * Deserialization schema from Maxwell JSON to Flink Table/SQL internal data structure {@link RowData}.
+ * The deserialization schema knows Maxwell's schema definition and can extract the database data
+ * and convert into {@link RowData} with {@link RowKind}.
+ *
+ * <p>Deserializes a <code>byte[]</code> message as a JSON object and reads
+ * the specified fields.
+ *
+ * <p>Failures during deserialization are forwarded as wrapped IOExceptions.
+ *
+ * @see <a href="http://maxwells-daemon.io/">Maxwell</a>
+ */
+public class MaxwellJsonDeserializationSchema implements DeserializationSchema<RowData> {
+	private static final long serialVersionUID = 1L;
+
+	private static final String OP_INSERT = "insert";
+	private static final String OP_UPDATE = "update";
+	private static final String OP_DELETE = "delete";
+
+	/** The deserializer to deserialize Maxwell JSON data. */
+	private final JsonRowDataDeserializationSchema jsonDeserializer;
+
+	/** TypeInformation of the produced {@link RowData}. **/
+	private final TypeInformation<RowData> resultTypeInfo;
+
+	/** Flag indicating whether to ignore invalid fields/rows (default: throw an exception). */
+	private final boolean ignoreParseErrors;
+
+	/** Number of fields. */
+	private final int fieldCount;
+
+	public MaxwellJsonDeserializationSchema(RowType rowType,
+											TypeInformation<RowData> resultTypeInfo,
+											boolean ignoreParseErrors,
+											TimestampFormat timestampFormatOption) {
+		this.resultTypeInfo = resultTypeInfo;
+		this.ignoreParseErrors = ignoreParseErrors;
+		this.fieldCount = rowType.getFieldCount();
+		this.jsonDeserializer = new JsonRowDataDeserializationSchema(
+			createJsonRowType(fromLogicalToDataType(rowType)),
+			// the result type is never used, so it's fine to pass in Canal's result type
+			resultTypeInfo,
+			false, // ignoreParseErrors already contains the functionality of failOnMissingField
+			ignoreParseErrors,
+			timestampFormatOption);
+	}
+
+	@Override
+	public RowData deserialize(byte[] message) throws IOException {
+		throw new RuntimeException(
+			"Please invoke DeserializationSchema#deserialize(byte[], Collector<RowData>) instead.");
+	}
+
+	@Override
+	public void deserialize(byte[] message, Collector<RowData> out) throws IOException {
+		try {
+			RowData row = jsonDeserializer.deserialize(message);
+			String type = row.getString(2).toString(); // "type" field
+			if (OP_INSERT.equals(type)) {
+				// difference with canal-json is that the "data" section in canal-json contains Array <Row <>>,
+				// and maxwell-json is Row <>
+				// So just take out Row<> in row and collect it
```

Comment:
We don't need to compare with canal-json in the comment. From my point of view, canal-json is a special case in encoding. You can just mention the "data" is the new image of insertion.